task/README.md

```diff
@@ -0,0 +1,243 @@
+| `secret`  | Boolean               |                                           | No       | `False`                       |
+| `maxgas`  | Integer               | Maximum gas for the invocation            | No       | 1000 <!-- ...or something --> |
+
+## 3.1 `type`
+
+The `type` field declares this object to be an IPVM Wasm configuration. The value MUST be `ipvm/wasm`.
+
+## 3.2 `with`
+
+The `with` field declares the Wasm module to load via CID.
+
+Note that the 
+
+## 3.3 `input`
+
+## 3.4 `maxgas`
+
+The `maxgas` field specifies the upper limit in gas that this task may consume.
+
+For the gas schedule, please see the [gas schedule spec].
+
+# 4 Effects
+
+The contract for effects is different from pure computation. As effects by definition interact with the "real world". These may be either commands or queries. Exmaples of effects include reading from DNS, sending an HTTP POST request, running a WASI module with network access, or receieving a random value.
+
+The `with` field MAY be filled from a relative value (previous step)
+
+| Field     | Type            | Description                         | Required | Default |
+|-----------|-----------------|-------------------------------------|----------|---------|
+| `type`    | `"ipvm/effect"` | Identify this workflow as an effect | Yes      |         |
+| `version` | SemVer          | IPVM effect schema version          | No       | `0.1.0` |
+| `using`   | URI             |                                     | Yes      |         |
+| `do`      | ability         |                                     | Yes      |         |
+| `args`    | `[{}]`          |                                     | No       | `[]`    |
+| `timeout` | Integer         | Timeout in milliseconds             | No       | `5000`  |
+| `auth`    | `[&UCAN]`       |                                     | No       | `[]`    |
+
+<!-- | `destructive` | Boolean         |  FIXME infer from `do` field?  | No       | `True`  | -->
+<!-- indeed these need to be registered by the runner -->
+
+``` json
+{
+    "type": "ipvm/effect",
+    "to": "did:key:zStEZpzSMtTt9k2vszgvCwF4fLQQSyA15W5AQ4z3AR6Bx4eFJ5crJFbuGxKmbma4",
+    "do": "crypto/sign",
+    "args": [
+        { "value": { "from": "earlierStep" } }
+    ]
+}
+```
+
+## 4.1 `type`
+
+
+``` json
+{
+  "type": "ipvm/task",
+  "version": "0.1.0",
+  "using": "wasm:Qm12345"
+  "args": {
+    "fun": "add_one",
+    "args": [1, 2, 3],
+    "maxgas": 1024
+  }
+}
+```
+
+``` json
+{
+  "type": "ipvm/task",
+  "version": "0.1.0",
+  "using": "docker:Qm12345"
+  "meta": {
+    "annotations": []
+  },
+  "args": {
+    "resources": {
+      "ram": {"gb": 10}
+    },
+    "inputs": [1, 2, 3],
+    "entry": "/",
+    "workdir": "/",
+    "env": {
+      "$FOO": "bar"
+    },
+    "timeout": {"seconds": "3600"},
+    "contexts": [],
+    "output": [],
+    "sharding": 5
+  }
+}
```

Comment:
@lukemarsden & @simonwo: this is a very rough mockup, but is this more in line with what you need?

Changes:

* They all just "tasks" (wasm, docker, etc)
*  Nested fields for docker-specific things
* Signalling Docker with `docker:`... not sure if there's an offical URI scheme from Docker/CNCF/OCI
* Annotations are a good idea! Here they're in the common fields section (un-nested)

Reply:
this looks broadly good!
* what determines the fields that are valid under `args`? is it the scheme in the `using` URI? I'd maybe prefer to have it be an explicit `kind` sub-field (open to a better name!) i.e. `type` is top-level `task`, `kind` is the kind of task it is, which determines the valid `args` fields. WDYT?

Reply:
Yeah, I was thinking of it as being scoped to the URI, but you're right that we probably want at least a version in there or something. We could also keep the same pattern of the `type` field inside the payload:

```js
{
  "type": "ipvm/task",
  "version": "0.1.0",
  "using": "docker:Qm12345"
  "meta": {
    "annotations": []
  },
  "args": {
    "type": "bacalhau/docker", // This line
    "version": "0.1.0", // Possibly this, too
    "resources": {
      "ram": {"gb": 10}
    },
    "inputs": [1, 2, 3],
    "entry": "/",
    "workdir": "/",
    "env": {
      "$FOO": "bar"
    },
    "timeout": {"seconds": "3600"},
    "contexts": [],
    "output": [],
    "sharding": 5
  }
}
```

Reply:
Yes!

Reply:
Also ideally we don't have bacalhau specific things, but generic things that bacalhau can implement. So we try and spearhead ipvm/wasm and add ipvm/docker